lib/AST/IRToASTVisitor.cpp

```diff
@@ -162,8 +162,20 @@ clang::Expr *IRToASTVisitor::CreateLiteralExpr(llvm::Constant *constant) {
       if (llvm::isa<llvm::ConstantInt>(constant)) {
         auto ci{llvm::cast<llvm::ConstantInt>(constant)};
         auto val{ci->getValue()};
-        if (val.getBitWidth() == 1U) {
+        auto val_bitwidth{val.getBitWidth()};
+        auto ull_bitwidth{ast_ctx.getIntWidth(ast_ctx.LongLongTy)};
+        if (val_bitwidth == 1U) {
+          // Booleans
           result = ast.CreateIntLit(val);
+        } else if (val_bitwidth > ull_bitwidth) {
```

Comment:
Are you able to check if the actual effective width is <= 64 (there is an `APInt` API for figuring out how many bits are needed to represent the value) and if so, turn it into `(uint128_t) constant`? Possible you can try to also detect if it is a sign-extended 64-bit value, or a `-1` or a `~0`, and then finally, if all else fails, I think the equation you want is:

```
(((uint128_t) high_64) << 64u) | low_64
```

Reply:
> Are you able to check if the actual effective width is <= 64 (there is an APInt API for figuring out how many bits are needed to represent the value) and if so, turn it into (uint128_t) constant?

I think we can do this. I actually thought we already do.

> Possible you can try to also detect if it is a sign-extended 64-bit value, or a -1 or a ~0

I could try to find bit patterns, but that would mean I'd be chasing special cases. I'd rather have a general solution here and maybe do a rewriting pass later.

> and then finally, if all else fails, I think the equation you want is ...

Yeah, on second thought this is going to be better.  I thought the `udiv` and `urem` would be more universal, but it only really scales towards `i128` anyway.